### DISCUSSION TITLE
About Bitmaps

### DISCUSSION THREAD

**Discussion by H1manshu21** [Blaze-Stars/Image-Filter-Application#19] [Q&A] [answered]
Learn about bitmaps and their usage.
Read: https://en.wikipedia.org/wiki/BMP_file_format
Watch: https://www.youtube.com/watch?v=0KmimFoalTI

Any other sources are welcome to discuss.

**Comment by Ank1taS** [comment, marked as answer]
A `bitmap` is a map of bits i.e bitmap is a 2D array of pixels where the image is an array of rows and each row is the array of pixels.

In an image like BMP (short for bitmap), JPEG, PNG, etc each pixel has some value according to image size that results in a specific color. BMP supports 1, 4, 8, 16, 24, and 32 bits color. Each pixel has 3 channels i.e `Red`, `Green` and `Blue`. 

A **8-bit BMP image means each pixel requires 8 bits size** and each pixel can have a maximum of **256 (2^8)** possible colors. Similarly, for a **16 bits and 24 bits image maximum of 65,536 and 16.7 million possible shades of color** can be represented in a pixel. Refer [this](https://docs.microsoft.com/en-us/dotnet/desktop/winforms/advanced/types-of-bitmaps?view=netframeworkdesktop-4.8)
therefore a **24 bits color refers to 8-bits red + 8-bits green + 8-bits blue** 

For a particular [height x weight], if we make pixel smaller then more numbers of pixels can fit into the screen and this results in a higher density of pixels (numbers of pixels per square inch).

Because more numbers bits are used to represent the color of each pixel and more numbers of pixels are placed so the image is of higher resolution.

BMP is a standard format used by Windows to store device-independent and application-independent images. The number of bits per pixel for a given BMP file is specified in a file header. BMP files with 24 bits per pixel are common. BMP files are usually not compressed and, therefore, are not well suited for transfer across the Internet.
BMP file also contains some **metadata** information like an image’s **height and width**. That metadata is stored at the beginning of the file in the form of two data structures generally referred to as `headers`

**Comment by Ank1taS** [comment]
#### Illustration of pixel color according to the values of the channel
Here is an example of an **8-bits image** When all 3 channels have different values and the pixel results in a particular shade.

1. When all the **3 channels values are 0**, it results in black as follows!
<img src="https://user-images.githubusercontent.com/65461370/169523406-4a3b6b74-b62c-4590-a4e7-6f187fdd9977.png" height="300px" width="500px">


2. When  **the Red channel has a smaller value and the rest 2 values are 0**, it results in a more red shade as follows!
<img src="https://user-images.githubusercontent.com/65461370/170857648-5479d60f-b7af-4d34-9252-f49e961ede1b.png" height="300px" width="500px">


3. When  **the Red channel has a higher value and the rest 2 values are 0**, it results in a more red shade as follows!

<img src="https://user-images.githubusercontent.com/65461370/169526909-0c1b0db6-e8dd-41ac-b8a0-7f5bb0ac875b.png" height="300px" width="500px">


4. When the **Red channel has the highest value and the rest 2 values are 0**, it results in a maximum red shade as follows!

<img src="https://user-images.githubusercontent.com/65461370/169527265-18258059-c208-4272-8bae-d4d56346c0dd.png" height="300px" width="500px">


5. When the **Green channel has the highest value and the rest 2 values are 0**, it results in a higher green shade as follows!

<img src="https://user-images.githubusercontent.com/65461370/169527617-13b26f30-9f62-4b09-a355-b16848ce73a0.png" height="300px" width="500px">


6. When the **Red channel has the highest value, the Green channel has 0 and Blue has a small value**, it a shade as follows!

<img src="https://user-images.githubusercontent.com/65461370/169527906-dc616e02-0e5f-4478-96bf-367f9147403f.png" height="300px" width="500px">


7. When the **Red channel has a small value, the Green channel has a higher value and the blue channel has comparatively little less value than Green**, it results in a maximum red shade as follows!

<img src="https://user-images.githubusercontent.com/65461370/169528524-663fc069-552f-4412-90a1-2e57800a1ae1.png" height="300px" width="500px">


8. When all the **channel has the highest value**, it results in white color as follows!

<img src="https://user-images.githubusercontent.com/65461370/169528691-6345c246-bf3f-44fd-aec0-8b66e8822e8f.png" height="300px" width="500px">

**Comment by chandrakant100** [comment]
@sabyasachi07 @Tushar7-coder @HembramBeta777 
Its a very Imortant dicussion since we will be working on this project using `bitmaps`

**Reply by SHembram18** [reply]
**Completed**

**Comment by Ank1taS** [comment]
### Factors affecting image quality

- The higher the resolution, the better the quality.

- The higher the color depth, the better the quality.

- The better the quality, the more memory required.

- Metadata (data about data) means image files are slightly bigger.

**Reply by Tushar7-coder** [reply]
As the above information here are the short defination of  few terms . 
1--> Resolution-pixel density
2-->Meta data-Data of data( Gives information about another data like size of image,height,width,name,location etc) .
